frontends/web/src/utils/event-legacy.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2021 Shift Crypto AG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { apiWebsocket } from '../utils/websocket';
+
+/**
+ * This type describes the subject of an event.
+ */
+
+type Subject = string;
+
+/**
+ * This interface models the legacy events that are received from the backend.
+ */
+
+export interface IEvent {
+    readonly data?: Subject;
+    readonly deviceID?: string;
+    readonly meta?: null;
+    readonly type?: string;
+    readonly code?: string;
+}
+
+/**
+ * This type describes the function used to observe the events.
+ */
+
+ type Observer = (event: IEvent) => void;
+
+/**
+ * This type describes the method returned to unsubscribe again.
+ */
+
+export type IUnsubscribe = () => void;
+
+/**
+ * This interface describes how the subscriptions are stored.
+ */
+
+interface ISubscriptions {
+    [subject: string]: Observer[];
+}
+
+/**
+ * Stores the subscriptions as an object-based hash map.
+ */
+
+const subscriptions: ISubscriptions = {};
+
+/**
+ * This function dispatches the events from the websocket to the observers.
+ */
+
+function handleMessages(data: IEvent): void {
+    if (
+        data.data
+        && typeof data.data === 'string'
+        && data.data in subscriptions
+        && subscriptions[data.data].length
+    ) {
+        // data.data in subscriptions && subscriptions[data.data].length
+        for (const observer of subscriptions[data.data]) {
+            observer(data);
+        }
+    }
+}
+
+/**
+ * Subscribes the given observer on events of the given subject and returns a method to unsubscribe.
+ */
+
+export function subscribe(subject: Subject, observer: Observer): IUnsubscribe {
+    if (!subscriptions[subject]) {
+        subscriptions[subject] = [];
+    }
+    const observers = subscriptions[subject];
+    if (observers.includes(observer)) {
+        console.warn(`observer already registered for ${subject}`);
+    }
+    observers.push(observer);
+    return () => {
+        if (!observers.includes(observer)) {
+            console.warn('!observers.includes(observer)');
+        }
+        const index = observers.indexOf(observer);
+        observers.splice(index, 1);
+        if (observers.includes(observer)) {
```

Comment:
should you better do it in a loop until `observers.includes(observer)` is false?

Reply:
No I'd prefer no loop if not necessary, includes is just the new indexOf > -1, why would you want to loop?

Reply:
why loop? but then why checking `if (observers.includes(observer)` after `observers.splice(index, 1)`?
either it's possible, in which case I guess you'll need to remove all instances in a loop, or it isn't. in the latter case, this check `if (observers.includes(observer)` is redundant.

Reply:
I am not sure I understand what you mean, the `subscribe` has to return a callback so that a react component can unsubscribe and stop listening.

This is same concept as the non-legacy version here https://github.com/digitalbitbox/bitbox-wallet-app/blob/master/frontends/web/src/utils/event.ts#L78-L101

Reply:
here's what I mean:

![image](https://user-images.githubusercontent.com/25405/108769103-d40b1500-7558-11eb-900c-a72f596a6631.png)

I don't understand what is `if (observers.includes(observer)) { console.warn(...) }` for.

Reply:
splice is synchronous and mutates the array, so the check after is redundant, remove this one.

Reply:
Ok I'll remove this but keep the other 2 before. The other 2 can be useful for if a consumer accidentally subscribes the same fn again without unsubscribing.